airflow/jobs/scheduler_job.py

```diff
@@ -1483,10 +1481,40 @@ def _do_scheduling(self, session) -> int:
             session.expunge_all()
             # END: schedule TIs
 
-            # TODO[HA]: Do we need to call
-            # _change_state_for_tis_without_dagrun (2x) that we were before
-            # to tidy up manually tweaked TIs. Do we need to do it every
-            # time?
+            # TODO[HA]: Do we need to do it every time?
+            try:
+                self._change_state_for_tis_without_dagrun(
+                    old_states=[State.UP_FOR_RETRY],
+                    new_state=State.FAILED,
+                    session=session
+                )
+
+                self._change_state_for_tis_without_dagrun(
+                    old_states=[State.QUEUED,
+                                State.SCHEDULED,
+                                State.UP_FOR_RESCHEDULE,
+                                State.SENSING],
+                    new_state=State.NONE,
+                    session=session
+                )
+
+                guard.commit()
+            except OperationalError as e:
+                # DB specific error codes:
+                # Postgres: 55P03
+                # MySQL: 3572, 'Statement aborted because lock(s) could not be acquired immediately and NOWAIT
+                #               is set.'
+                # MySQL: 1205, 'Lock wait timeout exceeded; try restarting transaction
+                #              (when NOWAIT isn't available)
+                db_err_code = getattr(e.orig, 'pgcode', None) or e.orig.args[0]
+
+                # We could test if e.orig is an instance of
+                # psycopg2.errors.LockNotAvailable/_mysql_exceptions.OperationalError, but that involves
+                # importing it. This doesn't
+                if db_err_code in ('55P03', 1205, 3572):
+                    self.log.debug("Lock held by another Scheduler")
+                else:
+                    raise
```

Comment:
Haven't added `session.rollback` here

Reply:
WDYT? We can add it before `raise` in the `else` block but might not need in `if` statement

Reply:
I think you should encapsulate this in an `is_lock_not_available_error` function to start with -- it makes it much clearer.

We'll need a rollback in the `if` success block, otherwise the connection isn't usable for the next steps.

Reply:
Updated